### ISSUE TITLE
PyTorch Distributed Elastic Launch Segmentation Fault with Python 3.12

### ISSUE THREAD

**Issue by PaulZhang12** [pytorch/pytorch#116423]
### 🐛 Describe the bug

With Python 3.12, using torch.distributed elastic_launch results in segmentation fault. Python 3.11 with the same code works.
```
import torch.distributed.launcher as pet
import uuid
import tempfile
import os


def get_launch_config(world_size: int, rdzv_endpoint: str) -> pet.LaunchConfig:
    return pet.LaunchConfig(
        min_nodes=1,
        max_nodes=1,
        nproc_per_node=world_size,
        run_id=str(uuid.uuid4()),
        rdzv_backend="c10d",
        rdzv_endpoint=rdzv_endpoint,
        rdzv_configs={"store_type": "file"},
        start_method="spawn",
        monitor_interval=1,
        max_restarts=0,
    )

def entrypoint():
    print("Here!")


if __name__ == '__main__':
    with tempfile.TemporaryDirectory() as tmpdir:
        launch_config = get_launch_config(2, os.path.join(tmpdir, "rdzv"))
        pet.elastic_launch(launch_config, entrypoint=entrypoint)()
```

### Versions

[conda] numpy                     1.26.2                   pypi_0    pypi
[conda] torch                     2.3.0.dev20231218+cu121          pypi_0    pypi
[conda] torchmetrics              1.0.3                    pypi_0    pypi
[conda] torchrec                  0.5.0.dev20231218+cu121          pypi_0    pypi

**Comment by malfet** [comment]
Can you attach a backtrace please?

**Comment by PaulZhang12** [comment]
Running the script does not give a stack trace. This simple repro is based off of torchrec metric tests that fail with Python 3.12, to which there is a stack trace with the seg fault: 
<img width="1700" alt="Screenshot 2023-12-27 at 6 04 23 PM" src="https://github.com/pytorch/pytorch/assets/28608314/68ef7f3e-f861-4d54-9dec-64d7e40b7408">

Also attaching the backtrace obtained when running the script with gdb: 
<img width="1700" alt="Screenshot 2023-12-27 at 6 03 40 PM" src="https://github.com/pytorch/pytorch/assets/28608314/4a54a2ca-7aab-4798-ab22-b11be56169bb">

**Comment by XilunWu** [comment]
@kurman The stacktrace shows segfault is raised when calling `store.compare_set()`. Not sure if it has anything to do with my libuv PR #116141 but our tests showed no failure IIRC. Not necessarily issue on our side but it's a heads-up.

**Comment by kurman** [comment]
> Not sure if it has anything to do with my libuv PR

@XilunWu I believe `useLibUV` false by default so this should not be a root cause. I am testing this now but I take that we should triage this under general python 3.12 support.

**Comment by wconstab** [comment]
Did you narrow this down to a specific root cause? @XilunWu

**Comment by wconstab** [comment]
Any update on this? 

It would be nice to know if the segfault is happening on the client or server side, and also whether it persists when we use the LibUV backend.  We plan to roll out libuv as default anyway, so that would be an easy fix if it were the case.

**Comment by c-p-i-o** [comment]
> whether it persists when we use the LibUV backend.

Issue persists even with LibUV. See #125990 for additional easier repro steps.
